### PR TITLE
[FIX] compare the sum of all quantity by product in lines between tranfer and invoice TASK#4914 HU#959 CA#2

### PR DESCRIPTION
**Pull request by hugho-ad**
**[WIP]**
- [x] Verify qty and product bewtween each transfer & invoice lines

FIX https://github.com/Vauxoo/yoytec/issues/500

**DUMMY https://github.com/Vauxoo/yoytec/pull/1068**

### CODE REVIEW

**Comment by hugho-ad** [comment]
@JulioSerna 
Can you review please.

I'm open to include any verifications you considers necessary

Regards

File: stock_picking_invoice_validation/model/stock_transfer_details.py [hunk 1]
```diff
@@ -51,25 +51,27 @@ def do_detailed_transfer(self):
...
+                for item in invoice.invoice_line:
+                    qty = inv_dict.get(item.product_id.id, 0) + item.quantity
+                    inv_dict.update({item.product_id.id: qty})
+                if inv_dict == tran_dict:
+                    transfer.picking_id.invoice_id = invoice.id
+                    res = super(
+                        StockTansferDetails, self).do_detailed_transfer()
+                    return res
                 else:
                     raise exceptions.Warning(
```
Comment (JulioSerna) [inline stock_picking_invoice_validation/model/stock_transfer_details.py]: @hugho-ad 
what happen if is partial delivery?,
e.g invoice line 10 quantity transfer item 1 quantity

Comment (hugho-ad) [inline stock_picking_invoice_validation/model/stock_transfer_details.py]: That transfer process will be done until quantities of each product in picking out, are completely availables in order to complete delivery based on invoice lines.

In other words, if a picking out are incompleate, we can't transfer to the customer.

I requested more details in HU 959, Until the response, the above are my thought.

Thanks for you support.
Regards!

**Comment by hugho-ad** [comment]
@JulioSerna can you help me to merge this please.

File: stock_picking_invoice_validation/model/sale_order.py [hunk 2]
```diff
@@ -8,17 +8,24 @@
...
 class SaleOrder(models.Model):
     _inherit = 'sale.order'
 
+    @api.multi
+    def _get_wh_conf(self):
+        key = "stock.check_inv_pick"
+        check_inv_pick = self.env["ir.config_parameter"].get_param(
+            key)
+        return check_inv_pick
+
```
Comment (nhomar) [inline stock_picking_invoice_validation/model/sale_order.py:21]: Hello.

Functional checks like this one can not be in ir.config.parameter, they must have its entries in config setting in the warehous section, please fix that.

**Comment by nhomar** [comment]
@hugho-ad I was waiting this module has a config no a ir.config.parameter.
 Show all checks

**Comment by hugho-ad** [comment]
@nhomar itś has in warehouse section:
image

---

**Comment by nhomar** [comment]
@hugho-ad I mean that the parameter must be saved in other place (human readable) not in the ir.config-parameter object.

the company is a good place.

File: stock_picking_invoice_validation/model/sale_order.py [hunk 3]
```diff
@@ -8,26 +8,27 @@
...
 
 
 class SaleOrder(models.Model):
     _inherit = 'sale.order'
 
+    @api.multi
+    def _get_wh_conf(self):
+        key = "stock.check_inv_pick"
+        check_inv_pick = self.env["ir.config_parameter"].get_param(key)
+        return check_inv_pick
```
Comment (nhomar) [inline stock_picking_invoice_validation/model/sale_order.py:20]: read this from the company.

**Comment by nhomar** [comment]
Is this ready @hugho-ad , rebase please.

**Comment by hugho-ad** [comment]
rebase Done, merge it please

**Comment by hugho-ad** [comment]
@nhomar Rebase Done again, 
After mege this
Mege this :https://github.com/Vauxoo/yoytec/pull/1067 (TODO: Remove dummy commit)

File: stock_picking_invoice_validation/model/stock_transfer_details.py [hunk 4]
```diff
@@ -50,26 +46,29 @@ def do_detailed_transfer(self):
...
+                for item in invoice.invoice_line.filtered(
+                        lambda r: r.product_id.type in ('consu', 'product')):
+                    qty = inv_dict.get(item.product_id.id, 0) + item.quantity
+                    inv_dict.update({item.product_id.id: qty})
+                if inv_dict == tran_dict:
+                    transfer.picking_id.invoice_id = invoice.id
+                    res = super(
+                        StockTansferDetails, self).do_detailed_transfer()
+                    return res
                 else:
```
Comment (nhomar) [inline stock_picking_invoice_validation/model/stock_transfer_details.py]: Please No el... return early.

File: stock_picking_invoice_validation/model/stock_transfer_details.py [hunk 5]
```diff
@@ -50,26 +46,29 @@ def do_detailed_transfer(self):
...
                     raise exceptions.Warning(
-                        _('Warning!'), _('Incorrect Invoice'))
+                        _('Warning!'),
+                        _('Incorrect Invoice, '
+                          'products and quantities are different '
+                          'between moves and invoice lines.'))
+                    # raise exceptions.Warning(
+                    #     _('Warning!'), _('Incorrect Invoice'))
             else:
```
Comment (nhomar) [inline stock_picking_invoice_validation/model/stock_transfer_details.py]: Again... No else... return early.

**Comment by nhomar** [comment]
Minimal comments in algorithm.. avoid the else returning earliest as commented.

**Comment by hugho-ad** [comment]
@nhomar comments done, and CI green 

Dummy : https://github.com/Vauxoo/yoytec/pull/1067

**Pull request merged**